var/spack/repos/builtin/packages/elsi/package.py

```diff
@@ -33,27 +33,58 @@ class Elsi(CMakePackage):
     variant("enable_sips", default=False, description="Enable SLEPc-SIPs support")
     variant("use_external_elpa", default=False, description="Build ELPA using SPACK")
     variant("use_external_ntpoly", default=False, description="Build NTPoly using SPACK")
-    variant("use_external_omm", default=False, description="Use external libOMM and MatrixSwitch")
     variant("use_external_superlu", default=False, description="Use external SuperLU DIST")
     variant(
         "use_mpi_iallgather", default=True, description="Use non-blocking collective MPI functions"
     )
+    variant("tests", default=False, description="Enable Fortran tests")
+    variant("tests_c", default=False, description="Enable C tests")
+    variant(
+        "internal_elpa_version",
+        default="2024",
+        values=("2024", "2023_11", "2023", "2021", "2020"),
+        description="Internal ELPA version",
+        multi=False,
+    )
```

Comment:
Does this choose a bundled ELPA version? IMO it would be better to default the `use_external_x` variants to true, or even remove them and always use the external libraries. I know this is a breaking change, but using the internal bundled versions is a guaranteed road to hell. There's precedent for removing the option to use bundled dependencies in https://github.com/spack/spack/pull/43210. This would also make it easier to model version constraints through spack's regular mechanisms instead of having to do it manually through the version variant.

I know this is a more intrusive change so I'm not expecting this to be changed now or ever, but as far as I could tell at least no other package currently requests explicitly any of the `~/+use_external_x` variants. This doesn't mean someone isn't using the variant in their environments of course...

Reply:
Yes, this variant is to choose the bundled ELPA version. The default is unfortunately still 2020.

I totally agree with your suggestion in principle, but ELPA is currently broken in Spack: https://github.com/spack/spack/issues/43902. Bundled libraries are non-optional in ELSI, so it is currently not possible to build cleanly with Spack using ELPA as an external dependency.

Reply:
Yep, fair enough, whatever it takes to get it working at this point.

What do you mean by

> Bundled libraries are non-optional in ELSI

though? The options are clearly there, so do you mean the options are broken or something similar? I'm not sure I see the relation to https://github.com/spack/spack/issues/43902 (if there is a relation...)?

Reply:
Yep, fair enough, whatever it takes to get it working at this point.

What do you mean by

> Bundled libraries are non-optional in ELSI

though? The options are clearly there, so do you mean the options are broken or something similar? I'm not sure I see the relation between bundled libraries and https://github.com/spack/spack/issues/43902 (if there is a relation...)?

Reply:
Sorry, I was unclear. By non-optional I mean that you can't just say `elsi~elpa`. ELPA is bundled and has to be installed, either from the bundled versions or as an external package. So if you remove the possibility to use the bundled versions, external ELPA is always required.

Reply:
Ok, that makes sense.

> So if you remove the possibility to use the bundled versions, external ELPA is always required.

This sounds like a win, but that you already know :sweat_smile: 

Looking at the ELSI CMake configuration, 2020 might be the default ELPA version, but it also looks like it only supports 2021 as an alternative version: https://github.com/ElectronicStructureLibrary/elsi-interface/blob/e27b7c925f30bc4486340e3cb6a32ffa4adc3de8/CMakeLists.txt#L103-L107. Or is there a branch somewhere that supports any version of ELPA when using the bundled version?

To summarize, my preference would be to completely remove the option of using bundled dependencies but I know this is drastic and I'm fine with not doing that. However, with the existing variants I would very much prefer to make the `use_external_x` variants default to false. This way users can also choose whatever version of ELPA they want (assuming it builds and is compatible...).

The above is also assuming I'm still not missing some weird corner case that requires preferring the bundled dependencies (e.g. ELSI not building with external ELPA).

Reply:
> Or is there a branch somewhere that supports any version of ELPA when using the bundled version?

The official project is hosted on GitLab: https://gitlab.com/elsi_project I think the GitHub project is a mirror, and it has not been updated in 2 years.

> To summarize, my preference would be to completely remove the option of using bundled dependencies but I know this is drastic and I'm fine with not doing that. However, with the existing variants I would very much prefer to make the use_external_x variants default to false.

I agree, but to do this we need a working `elpa+cuda` in Spack.

I was hoping that who broke the package would fix it, but despite multiple pings from my part this is not the case. At this point it looks like I will have to fix it myself...

Reply:
Thanks @RMeli for explaining this offline. The issue is that the external ELPA built through spack with CUDA does not build, whereas the internal bundled ELPA somehow does... So preferring the internal bundled ELPA indeed seems like a requirement for now (and may warrant a conflict?). Other than a possible addition of a conflict you can consider this thread resolved.

Reply:
A glimps of hope from https://github.com/spack/spack/pull/44749, I'll look into using only external packages if they build well.